src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs

```diff
@@ -292,19 +256,43 @@ private void WatchForFileSystemEventsThreadStart()
                     // If this event is the second in a rename pair then skip it
                     continue;
                 }
-                else if ((DoesPathPassNameFilter(eventPaths[i])) && ((_filterFlags & eventFlags[i]) != 0))
+                else if ((DoesPathPassNameFilter(path)) && ((_filterFlags & eventFlags[i]) != 0))
                 {
-                    // Remove the base directory prefix (including trailing / that OS X adds)
-                    string relativePath = eventPaths[i].Remove(0, _fullDirectory.Length + 1);
+                    // The base FileSystemWatcher does a match check against the relative path before combining with 
+                    // the root dir; however, null is special cased to signify the root dir, so check if we should use that.
+                    string relativePath = null;
+                    if (eventPaths[i].Equals(_fullDirectory, StringComparison.OrdinalIgnoreCase) == false)
+                    {
+                        // Check if the event path, with the root directory removed, begins with a / and
+                        // if so, remove it; otherwise, just remove the root path (which contains a trailing /)
+                        if (eventPaths[i][_fullDirectory.Length] == '/')
```

Comment:
@stephentoub Now that I think about it, should this be Path.DirectorySeparatorChar?